statistics/cmsketch.go

```diff
@@ -662,23 +662,40 @@ func MergeTopN(topNs []*TopN, n uint32) (*TopN, []TopNMeta) {
 	sort.Slice(sorted, func(i, j int) bool {
 		return sorted[i] > sorted[j]
 	})
-	n = mathutil.MinUint32(uint32(numTop), n)
+	remain := mathutil.MinUint32(uint32(numTop), n)
 	// lastTopCnt is the smallest value in the new TopN structure
 	var lastTopCnt uint64
-	if n > 0 {
-		lastTopCnt = sorted[n-1]
+	if remain > 0 {
+		lastTopCnt = sorted[remain-1]
 	}
 
 	var finalTopN TopN
-	finalTopN.TopN = make([]TopNMeta, 0, n)
-	popedTopNPair := make([]TopNMeta, 0, uint32(numTop)-n)
+	finalTopN.TopN = make([]TopNMeta, 0, remain)
+	popedTopNPair := make([]TopNMeta, 0, uint32(numTop)-remain)
+	boundaryTopN := make([]TopNMeta, 0, numTop)
 	for value, cnt := range counter {
 		data := hack.Slice(string(value))
```

Comment:
We can check `remain` first. For example, `if remain < 0 {break}`. And use `if / else if / else` to check the relation between `cnt` and `lastTopCnt`;